### PR TITLE
Fix #272: cherry pick api and helpers implemented

### LINKED ISSUE #272: Enable CherryPicking with Category filtering
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the COMET Web Services
- [x] I have searched [open](https://github.com/RHEAGROUP/COMET-WebServices-Community-Edition/issues) and [closed](https://github.com/RHEAGROUP/COMET-WebServices-Community-Edition/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description

We should be able to request Things based on special Category filtering (by ShortName) inside an Iteration.

The route will looks like: http://localhost:5000/EngineeringModel/{_EMId_}/Iteration/{_IterationId_}/?classkind=[{Values of classKind}]&category=[{Shortname of each Category}]

Server side, data should be retrieved inside the requested iteration based on the classkind. For now, most important things to retrieve is ElementDefinition/ElementUsage that are contained inside the Iteration and Parameter where the ParameterType are categorized by the specific Category.

### PR DESCRIPTION
**Pull request by antoineatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/COMET-WebServices-Community-Edition/pulls) open
- [x] I have verified that I am following the COMET-SDK [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/COMET-WebServices-Community-Edition/master/.github/CONTRIBUTING.md)
- [ ] I have provided test coverage for my change (where applicable)

### Description

Fix #272 

- CherryPick API EndPoint created
- Retrieve all contained things inside an Iteration
- Filter Things based on the ClassKind and on Category ShortNames

### CODE REVIEW

File: CDP4WebServices.API/Services/Protocol/IQueryParameters.cs [hunk 1]
```diff
@@ -71,6 +75,16 @@ public interface IQueryParameters
...
 
+        /// <summary>
+        /// Gets or sets a collection of <see cref="ClassKind"/> to used during the cherry picking request
+        /// </summary>
+        IEnumerable<ClassKind> ClassKinds { get; set; }
+
+        /// <summary>
+        /// Gets or sets a collection of <see cref="Category"/>s shortname to used during the cherry picking request
+        /// </summary>
+        IEnumerable<string> CategoriesShortName { get; set; }
```
Comment (samatrhea) [inline CDP4WebServices.API/Services/Protocol/IQueryParameters.cs:86]: i doubt using shortnames wil work, they cannot be guaranteed to be unique accross the chain of RDLs. Please refactor to use Guids, feel free to use shortguids

File: CDP4WebServices.API/Helpers/CherryPickHelper.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    using CDP4Common.CommonData;
+    using CDP4Common.DTO;
+    using CDP4Common.Extensions;
+
+    using Thing = CDP4Common.DTO.Thing;
+
+    /// <summary>
+    /// Helper for the cherry picking features
+    /// </summary>
+    public static class CherryPickHelper
```
Comment (samatrhea) [inline CDP4WebServices.API/Helpers/CherryPickHelper.cs:42]: please refactor to an injectable service.

File: CDP4WebServices.API/Helpers/ContainmentHelper.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    using CDP4Common.CommonData;
+    using CDP4Common.Extensions;
+
+    using Thing = CDP4Common.DTO.Thing;
+
+    /// <summary>
+    /// Helper class that provide capabilities on retrieving contained <see cref="Thing"/> or container <see cref="Thing"/>
+    /// </summary>
+    public static class ContainmentHelper
```
Comment (samatrhea) [inline CDP4WebServices.API/Helpers/ContainmentHelper.cs:40]: please refactor to injectable service

File: CDP4WebServices.API/Modules/10-25/CherryPickApi.cs [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+            var sw = new Stopwatch();
+            sw.Start();
+            var requestToken = this.GenerateRandomToken();
+
+            try
+            {
+                Logger.Info(this.ConstructLog($"{requestToken} started"));
+                HttpRequestHelper.ValidateSupportedQueryParameter(this.Request, this.RequestUtils, SupportedGetQueryParameters);
+
```
Comment (samatrhea) [inline CDP4WebServices.API/Modules/10-25/CherryPickApi.cs:129]: i think that we should verify that the caller has permissions to request the classkind before we process the request. If the classkind does not have at leas READ, we should exit here and return a 404 not allowed. I want to make sure we do not create a backdoor here.

**Comment by samatrhea** [comment]
please add unit tests to keep coverage to at least the current value or better. Core product, cannot accept code without coverage

**Review by lxatrhea — APPROVED** [review]
I agree with Sam's comments. The rest looks good to me.

**Comment by antoineatrhea** [comment]
@samatrhea Included your change requests and created unit tests for Services.

**Pull request merged**